Feature/wastewater survey processing

Bachibouzouk opened the pull request.

Bachibouzouk committed:
Add a method to add a single bus to the scenario datapackage

Bachibouzouk committed:
Add wastewater processing

Bachibouzouk commented on component_library/scripts/build_scenario.py:
```diff
@@ -87,6 +87,102 @@ def water_systems_postprocessing(self, survey):
+            self.components.update({("septic_system", "grey_water_septic"): {"water_in_bus": "", "water_out_bus": ""}})
+
+        elif "constructed wetland" in wastewater_systems:
+            if "constructed_wetland" in self.components:
+                self.components["constructed_wetland"].update({"constructed_wetland": {"name": "grey_water_cw"}})
+            else:
+                self.components.update({"constructed_wetland": {"name": "grey_water_cw"}})
```
```suggestion
            self.components.update({("constructed_wetland", "grey_water_cw"): {"water_in_bus": "", "water_out_bus": ""}})
```

Bachibouzouk replied:
@Vivek-H-Rana - you can fix the calls below according to this suggestion (you turn 4 lines into one)

Bachibouzouk commented on component_library/scripts/build_scenario.py:
```diff
@@ -87,6 +87,102 @@ def water_systems_postprocessing(self, survey):
+                        {"decentralized_waste_water_treatment_plant": {"name": "decentralized_WWTP"}})
+
+        # output waste water treatment plant water bus
+        self.components.update({"bus-op-wwtp-water": {"name": "op-wwtp-water-bus"}})
```
if you need a bus you can now use `self.add_single_bus(name=..., carrier=...)`

Bachibouzouk commented on component_library/scripts/build_scenario.py:
```diff
@@ -87,6 +87,102 @@ def water_systems_postprocessing(self, survey):
+
+        if "flush toilet" in toilet_types:
+            if "septic system" in wastewater_systems:
+                self.components.update({("septic_system", "black_water_septic"): {"water_in_bus": "", "water_out_bus": ""}})
```
The water in bus name and water out bus name must be provided (we need to add busses between all the components of the train

Vivek-H-Rana committed:
waste water systems postprocessing logic hardcode change

Bachibouzouk committed:
Adapted the component for toilets in the component library

Bachibouzouk committed:
Process question 7.3 automatically

Bachibouzouk committed:
Create a convenience method to add single components

Bachibouzouk committed:
Do not overwrite existing busses in add_busses

Bachibouzouk committed:
Fix default bus for septic system

Bachibouzouk committed:
Fix indent

MaaJoo13 committed:
Add component load.csv to the scenario, add the load profiles to profiles.csv

small fix if there is no demand_data

another small bug fix

Revert "another small bug fix"

This reverts commit 778fc8ddd79fba8b842d0a7b6c3f6f8d0a2a5fb0.

Vivek-H-Rana committed:
survey component mapping water treatment updated

Vivek-H-Rana committed:
wip components water treatement csv updated

Vivek-H-Rana committed:
water systems postprocessing survey access check

Vivek-H-Rana committed:
update survey mapping, update water systems postprocessing, improve process_survey function

Vivek-H-Rana committed:
add function to formulate drinking and service water component lists

Vivek-H-Rana committed:
add function to reorder drinking and service water component lists by main treatment train

Vivek-H-Rana committed:
foreign key fix in WIP component and scenairo 1 json

Vivek-H-Rana committed:
water in and out buses added to WIP component bus.csv

Vivek-H-Rana committed:
water in and out buses added to WIP component bus.csv

Vivek-H-Rana committed:
Merge branch 'feature/wastewater-survey-processing' of https://github.com/rl-institut/WEFEConfigurator into feature/wastewater-survey-processing

Vivek-H-Rana committed:
add function to create dictionary with water treatment train components and buses

Bachibouzouk committed:
All WIP changes except additional bus

commit c0e79ce0b81770ea87f231f4172c174e7c6908e8 (HEAD ->
feature/crops_survey_processing)
Author: Max Libberoth <148241644+MaaJoo13@users.noreply.github.com>
Date:   Mon Sep 22 15:03:49 2025 +0200

    WIP for PF to temper with

commit 2c55fbb3f0beb9b62051b37f4b15b0fdeb5c92ca
Author: Max Libberoth <148241644+MaaJoo13@users.noreply.github.com>
Date:   Fri Sep 19 13:45:50 2025 +0200

    WIP: additional busses for crops; crop and biomass excess

commit 2196128f8d4d9c86ea9da092f636b062374a3f5c
Author: Max Libberoth <148241644+MaaJoo13@users.noreply.github.com>
Date:   Thu Sep 18 17:13:51 2025 +0200

    WIP: add sources and sinks to crop components

commit b18d02f6849b67d61730b39954b46efe4e490c8c
Author: Max Libberoth <148241644+MaaJoo13@users.noreply.github.com>
Date:   Wed Sep 17 18:50:12 2025 +0200

    WIP: adapt survey_answer_component_mapping to crops

Bachibouzouk committed:
Consider additional busses separately

Bachibouzouk commented on component_library/scripts/build_scenario.py:
```diff
@@ -508,15 +509,19 @@ def update_component_attributes(self, component_key,component_params):
+        if not isinstance(component_attrs, dict):
             component_attrs = {}
+            logging.warning(f"The component attributes {component_attrs} must be of type dict! Will be ignored...")
 
         if component_name is None:
             component_key = (component_type, component_type)
-        elif not isinstance(component_key, tuple):
-            raise TypeError("The component key provided is neither a string not a tuple")
-        else:
+        elif isinstance(component_name, tuple):
+            component_key = component_name
+        elif isinstance(component_name, str):
             component_key = (component_type, component_name)
+        else:
+            logging.warning(f"The component name {component_name} is neither a string nor a tuple")
+
```
I am not sure why this change is needed

Vivek-H-Rana replied:
I don't think i ever touched this function at all, maybe Max changed it to accomodate some errors from his side

Vivek-H-Rana committed:
fix duplicate bus generation in bus.csv

Vivek-H-Rana committed:
fix bus duplication again and water components duplication try

Vivek-H-Rana committed:
fix water components duplication between process_survey and water_systems_postprocessing

MaaJoo13 committed:
Fix datapackage to correctly add load and corresponding excess

MaaJoo13 committed:
Fix datapackage to correctly add load and corresponding excess

MaaJoo13 committed:
Add components to component_lib, adapt existing components

MaaJoo13 committed:
Adapt survey-answer-component-mapping to new components/adapted components

Bachibouzouk committed:
Infer some foreign keys

MaaJoo13 committed:
WIP create running scenarios from survey answers

MaaJoo13 committed:
WIP supposedly correct scenario with all energy components

MaaJoo13 committed:
Add busses and profiles metadata to datapackage.json

MaaJoo13 committed:
fix scenario builder and componnet lib, running system with energy-only components

MaaJoo13 committed:
WIP working 'scenario_15' with energy-only components

MaaJoo13 committed:
Rebase everything

MaaJoo13 committed:
Merge branch 'feature/crops_survey_processing' into feature/wastewater-survey-processing

MaaJoo13 committed:
Combine drinking and service water if not differentiated

MaaJoo13 committed:
Add important survey answers as class attributes, restructure demand processing to include water storages, improve component library

MaaJoo13 committed:
Improve component library: Economic data for energy components, remove unused components

MaaJoo13 committed:
Restore profile metadata in newly generated datapackage.json, fix
component mapping

MaaJoo13 committed:
Adapt component lib (energy components) to pre_processing_costs/MOO, small fixes to ScenarioBuilder

Vivek-H-Rana committed:
Improve component library: Economic data for water treatment components

Vivek-H-Rana committed:
Improve component library: Economic data for toilet components

Vivek-H-Rana committed:
Improve component library: Economic data for wastewater treatment components

Vivek-H-Rana committed:
Fix water_consumption_factor type in toilets.csv

Vivek-H-Rana committed:
Update datapackage.json in WIP_components to include toilets, water, and wastewater treatment

Vivek-H-Rana committed:
Remove SWRO and Desalinator in survey answer component mapping

Vivek-H-Rana committed:
Update bus.csv in WIP_components to include toilets, water, and wastewater treatment

Vivek-H-Rana committed:
Improve component library: add sludge_out_bus in wastewater_treatment.csv

Vivek-H-Rana committed:
Adapt water and wastewater systems post-processing to handle automatic updates of capacity, efficiency, and specific energy consumption based on survey answers

Vivek-H-Rana committed:
Add normalized waste profiles for toilets and wastewater treatment components

Vivek-H-Rana committed:
Add waste sources and update their links in profiles.csv and foreign keys in datapackage for WIP components

Vivek-H-Rana committed:
Add excess in WIP components for toilets, water, and wastewater treatment

Vivek-H-Rana committed:
Add waste source and excess components for toilets, water, and wastewater treatment through survey_answer_component_mapping_in_use.json

Vivek-H-Rana committed:
Split toilets, water, and wastewater treatment CSVs in WIP components to accommodate optional buses

Vivek-H-Rana committed:
Update datapackage.json in WIP components to reflect the CSV split for toilets, water, and wastewater treatment

Vivek-H-Rana committed:
Add working scenario with diesel and drinking water: scenario_20_wt

JulianBarinton committed:
adding water-pump.csv, and addapting component mapping accordingly
updated data package
removed question regarding energy source of water pump for simplification

Vivek-H-Rana committed:
System infeasible toilet system

JulianBarinton committed:
adding excess-human-feces & excess-human-urine for avoiding infeasibility. A high "penalty" marginal cost is included to incentivize using the actual toilet

MaaJoo13 committed:
Fix water pump system integration, add missing water sources

MaaJoo13 committed:
Add water pipes if no pumps are used

MaaJoo13 committed:
Add new/adapted water sources to component mapping

Vivek-H-Rana committed:
Update waste profiles in waste_data.csv

MaaJoo13 committed:
Assign water pump attributes correctly

MaaJoo13 committed:
validate component lib

Vivek-H-Rana committed:
Update Scenarios: Removal of unwanted scenarios and Addition of Working Scenario 20

Vivek-H-Rana committed:
Fix build_scenario.py: correct water/wastewater postprocessing and waste data import

Vivek-H-Rana committed:
Add scenario 20 survey answers

MaaJoo13 committed:
Integrate cf_aware in add_sequences

Vivek-H-Rana committed:
Update WIP_components: Component CSVs and datapackage.json

Vivek-H-Rana committed:
Merge after pull

JulianBarinton committed:
Merge branch 'feature/wastewater-survey-processing' into feature/water-pump

JulianBarinton committed:
Merge pull request #18 from rl-institut/feature/water-pump

looks fine. Issue with power rating [kW] being used for calculating water throughput if not given can be fixed in a later commit (low priority for now)

MaaJoo13 committed:
Regenerate datapackage.json

MaaJoo13 committed:
WIP small fixes to build_scenario

Vivek-H-Rana committed:
Set input bus of water treatment trains to untreated-water-bus

Vivek-H-Rana committed:
Improve survey, survey component mapping and subquestion mapping

Vivek-H-Rana committed:
Fix models.py for null/empty values from survey

Vivek-H-Rana committed:
Fix for tech mismatch between pollutants selected and technology present in WT

Vivek-H-Rana committed:
Set output bus of water treatment trains to their respective water bus

Vivek-H-Rana committed:
Fix handling of user given technology parameters and null values in subquestion matrix

Vivek-H-Rana committed:
Fix handle excess capacity for drinking and service water treatment trains

Vivek-H-Rana committed:
Refactor(wastewater): improve toilets/WWTP post-processing in build_scenario.py

Vivek-H-Rana committed:
Update waste sources and excess in WIP components

Vivek-H-Rana committed:
Fix water sources addition in survey answer component mapping when distinction between drinking and service water

Vivek-H-Rana committed:
Link waste source capacities to gui population input

Vivek-H-Rana committed:
Working updates mirroring WEGEgui

Vivek-H-Rana committed:
Improve 'criteria_' stripping

Vivek-H-Rana committed:
Refactor water_systems_postprocessing to reduce duplication

Vivek-H-Rana committed:
Add annuity field to components and datapackage in component lib if missing

Vivek-H-Rana committed:
Add pre-, core-, and post-treatment categories to water treatment dictionary

Vivek-H-Rana committed:
Fix custom attributes and set skip_preprocessing to False in compute.py

Vivek-H-Rana committed:
Add flag-controlled water treatment simplification into pre-, core-, and post-treatment by modifying component csvs and datapackage in build_scenario.py

Vivek-H-Rana committed:
Fix service water handling in modify_bus_csv()

Vivek-H-Rana committed:
Fix scenario path handling by converting local paths to Path objects in compute.py

Vivek-H-Rana commented:
@MaaJoo13 we really should be merging this to dev/production branch, it has like 200 plus commits

Vivek-H-Rana committed:
Update WIP component library CSVs and datapackage.json to align with WEFEgui

Vivek-H-Rana committed:
Add country-specific component library updater

Load {country}_params.py and update matching columns in elements/*.csv files.
- Dynamic import of user params
- Robust value comparison (handles NaN/float/str/bool)
- Verbose logging of changes
- Graceful handling of missing files/columns/components

Usage: update_component_library(path, "kenya")

Vivek-H-Rana committed:
Add country parameter support to build_scenario.py

Vivek-H-Rana committed:
Add Generic country parameter template for WEFE component library updates

Vivek-H-Rana committed:
Add WEFEConfigurator units reference and link in generic_country_params.py

Vivek-H-Rana committed:
Fix aggregation logic for simplification in pre, core and post water treatment blocks
